ru/lessons/specifics/ecto.md

```diff
@@ -0,0 +1,342 @@
+
+```elixir
+query = from u in User,
+    where: fragment("downcase(?)", u.username) == ^username
+    select: u
+```
+
+Примеры построения различных запросов можно найти здесь [Ecto.Query.API](http://hexdocs.pm/ecto/Ecto.Query.API.html).
+
+## Изменяемые наборы
+
+В предыдущем параграфе мы увидели как извлечь необходимую информацию из базы данных, а что на счет добавления и обновления записей? Для этого нам необходимы изменяемые наборы (Changesets).
+
+Изменяемые наборы выполняют функции фильтрации, валидации, и поддержки ограничений при изменении модели.
+
+Давайте используем пример изменяемых наборов для создания аккаута для пользователя. Для начала необходимо обновить модель:
+
+```elixir
+defmodule ExampleApp.User do
+  use Ecto.Schema
+  import Ecto.Changeset
+  import Comeonin.Bcrypt, only: [hashpwsalt: 1]
+
+  schema "users" do
+    field :username, :string
+    field :encrypted_password, :string
+    field :email, :string
+    field :confirmed, :boolean, default: false
+    field :password, :string, virtual: true
+    field :password_confirmation, :string, virtual: true
+
+    timestamps
+  end
+
+  @required_fields ~w(username email password password_confirmation)
+  @optional_fields ~w()
+
+  def changeset(user, params \\ :empty) do
+    user
+    |> cast(params, @required_fields, @optional_fields)
+    |> validate_length(:password, min: 8)
+    |> validate_password_confirmation()
+    |> unique_constraint(:username, name: :email)
+    |> put_change(:encrypted_password, hashpwsalt(params[:password]))
+  end
+
+  defp validate_password_confirmation(changeset) do
+    case get_change(changeset, :password_confirmation) do
+      nil ->
+        password_mismatch_error(changeset)
+      confirmation ->
+        password = get_field(changeset, :password)
+        if confirmation == password, do: changeset, else: password_incorrect_error(changeset)
+    end
+  end
+
+  defp password_mismatch_error(changeset) do
+    add_error(changeset, :password_confirmation, "Passwords does not match")
+  end
+
+  defp password_incorrect_error(changeset) do
+    add_error(changeset, :password, "is not valid")
+  end
+end
+```
+
+Таким образом мы улучшили функцию `changeset/2`, а также добавили три новые вспомогательные функции: `validate_password_confirmation/1`, `password_mismatch_error/1`, и `password_incorrect_error/1`.
+
+По названию можно догадаться, что функция `changeset/2` создает новый изменяемый набор. В теле функции используем функцию `cast/4`, которая трансформирует данные параметры в изменяемый набор, который состоит из обязатаельных и необязательных полей. Далее происходит валидация длинны пароля, находящегося в изменяемом наборе. После этого, используя созданную нами фукнцию, мы можем проверить, что подтверждение пароля является таким же как и сам пароль, а также проверить уникальность имени пользователя. В конце, обновим пароль в базе данных. Для этого используем функцию `put_change/3`, которая обновит значение в именяемом наборе.
```

Comment:
o dear. thank you for noticing. I was getting a bit tired closer to the end of the article. :sleepy: